src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -0,0 +1,292 @@
+        private readonly PinnedRemotableDataScope _scope;
+        private readonly CancellationToken _cancellationToken;
+
+        public static async Task<SolutionAndSessionHolder> CreateAsync(RemoteHostClient.Session session, PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+        {
+            var sessionWithSolution = new SolutionAndSessionHolder(session, scope, cancellationToken);
+
+            try
+            {
+                await session.RegisterPinnedRemotableDataScopeAsync(scope).ConfigureAwait(false);
+                return sessionWithSolution;
+            }
+            catch
+            {
+                sessionWithSolution.Dispose();
+
+                // we only expect this to happen on cancellation. otherwise, rethrow
+                cancellationToken.ThrowIfCancellationRequested();
+                throw;
+            }
+        }
+
+        private SolutionAndSessionHolder(RemoteHostClient.Session session, PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+        {
+            _session = session;
+            _scope = scope;
+            _cancellationToken = cancellationToken;
+        }
+
+        public void AddAdditionalAssets(CustomAsset asset)
+        {
+            _scope.AddAdditionalAsset(asset, _cancellationToken);
+        }
+
+        public void Dispose()
+        {
+            _scope.Dispose();
+            _session.Dispose();
+        }
+
+        public Task InvokeAsync(string targetName, params object[] arguments) =>
+            _session.InvokeAsync(targetName, arguments);
+        public Task<T> InvokeAsync<T>(string targetName, params object[] arguments) =>
+            _session.InvokeAsync<T>(targetName, arguments);
+        public Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync(targetName, arguments, funcWithDirectStreamAsync);
+        public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync<T>(targetName, arguments, funcWithDirectStreamAsync);
+    }
+
+    /// <summary>
+    /// Helper type for common session case
+    /// 
+    /// This will let one to hold onto <see cref="RemoteHostClient.Session"/> for a while.
+    /// this helper will let you not care about remote host being gone while you hold onto the session if that ever happen
+    /// 
+    /// and also make sure state is correct even if multiple threads call TryInvokeAsync at the same time. but this 
+    /// is not optimized to handle highly concurrent usage. if highly concurrent usage is required, either using
+    /// <see cref="RemoteHostClient.Session"/> direclty or using <see cref="SolutionAndSessionHolder"/> would be better choice
+    /// </summary>
+    internal sealed class KeepAliveSessionHolder
```

Comment:
I don't want it to be session. I want session to be a session on pure communication. all others are just helper.